lucet-concurrency-tests/tests/killswitch.rs

```diff
@@ -0,0 +1,731 @@
+
+        let (kill_switch, outstanding_killswitch) = (inst.kill_switch(), inst.kill_switch());
+
+        let t = thread::Builder::new()
+            .name("guest".to_owned())
+            .spawn(move || {
+                match inst.run("infinite_loop", &[]) {
+                    Err(Error::RuntimeTerminated(TerminationDetails::Remote)) => {
+                        // this is what we want!
+                    }
+                    res => panic!("unexpected result: {:?}", res),
+                }
+
+                // A freshly acquired kill switch can cancel the next execution.
+                // Test here rather than the outer test body because this closure moves `inst`.
+                assert_eq!(inst.kill_switch().terminate(), Ok(KillSuccess::Cancelled));
+            })
+            .expect("can spawn a thread");
+
+        let terminator = in_guest.wait_and_then(move || {
+            thread::spawn(move || {
+                assert_eq!(kill_switch.terminate(), Ok(KillSuccess::Signalled));
+            })
+        });
+
+        t.join().unwrap();
+        terminator.join().unwrap();
+
+        // Outstanding kill switches fail, because the kill state was reset.
+        assert_eq!(outstanding_killswitch.terminate(), Err(KillError::Invalid));
+    })
+}
+
+#[test]
+fn terminate_entering_guest() {
+    let test_entering_guest_before_domain_change: fn(&Instance) -> SyncWaiter =
+        |inst: &Instance| -> SyncWaiter {
+            inst.lock_testpoints
+                .instance_entering_guest_before_domain_change
+                .wait_at()
+        };
+    let test_entering_guest_after_domain_change: fn(&Instance) -> SyncWaiter =
+        |inst: &Instance| -> SyncWaiter {
+            inst.lock_testpoints
+                .instance_entering_guest_after_domain_change
+                .wait_at()
+        };
+
+    for (i, racepoint_builder) in [
+        test_entering_guest_before_domain_change,
+        test_entering_guest_after_domain_change,
+    ]
+    .iter()
+    .enumerate()
+    {
+        println!("testing racepoint {}", i);
+        test_exclusive_instance_with_instrumented_guest_entry(|mut inst| {
+            let kill_switch = inst.kill_switch();
+            let racepoint = racepoint_builder(&inst);
+
+            let guest = thread::Builder::new()
+                .name("guest".to_owned())
+                .spawn(move || match inst.run("run_guest", &[]) {
+                    Err(Error::RuntimeTerminated(TerminationDetails::Remote)) => {}
+                    res => panic!("unexpectd result: {:?}", res),
+                })
+                .expect("can spawn thread to run guest");
+
+            racepoint.wait_and_then(|| {
+                kill_switch.terminate().expect("can terminate in guest");
+            });
+
+            guest.join().expect("guest exits without panic");
+        })
+    }
+}
+
+// Test a termination that completes right before `exit_guest_region` takes ownership of termination.
```

Comment:
I'm realizing that what would help me with the clarity of these tests is if the brief header comments described the test cases using the same terms or language as the docs.  I believe this the case where the KillSwitch is in a race with teardown and KillSwitch acquires terminable.  I don't feel confident in that assessment.

Reply:
Nope.  I'm wrong.  I think that's the test case on line 352.  Maybe?

Reply:
"link these tests with cases discussed in the documentation" is a good idea, and I've done just that.